variants/mpark/variant.hpp

```diff
@@ -0,0 +1,1954 @@
+  get_if(variant<Types...>*) noexcept;
+
+  template <size_t I, class... Types>
+  constexpr add_pointer_t<const variant_alternative_t<I, variant<Types...>>>
+  get_if(const variant<Types...>*) noexcept;
+
+  template <class T, class... Types>
+  constexpr add_pointer_t<T>
+  get_if(variant<Types...>*) noexcept;
+
+  template <class T, class... Types>
+  constexpr add_pointer_t<const T>
+  get_if(const variant<Types...>*) noexcept;
+
+  // 20.7.5, relational operators
+  template <class... Types>
+  constexpr bool operator==(const variant<Types...>&, const variant<Types...>&);
+
+  template <class... Types>
+  constexpr bool operator!=(const variant<Types...>&, const variant<Types...>&);
+
+  template <class... Types>
+  constexpr bool operator<(const variant<Types...>&, const variant<Types...>&);
+
+  template <class... Types>
+  constexpr bool operator>(const variant<Types...>&, const variant<Types...>&);
+
+  template <class... Types>
+  constexpr bool operator<=(const variant<Types...>&, const variant<Types...>&);
+
+  template <class... Types>
+  constexpr bool operator>=(const variant<Types...>&, const variant<Types...>&);
+
+  // 20.7.6, visitation
+  template <class Visitor, class... Variants>
+  constexpr see below visit(Visitor&&, Variants&&...);
+
+  // 20.7.7, class monostate
+  struct monostate;
+
+  // 20.7.8, monostate relational operators
+  constexpr bool operator<(monostate, monostate) noexcept;
+  constexpr bool operator>(monostate, monostate) noexcept;
+  constexpr bool operator<=(monostate, monostate) noexcept;
+  constexpr bool operator>=(monostate, monostate) noexcept;
+  constexpr bool operator==(monostate, monostate) noexcept;
+  constexpr bool operator!=(monostate, monostate) noexcept;
+
+  // 20.7.9, specialized algorithms
+  template <class... Types>
+  void swap(variant<Types...>&, variant<Types...>&) noexcept(see below);
+
+  // 20.7.10, class bad_variant_access
+  class bad_variant_access;
+
+  // 20.7.11, hash support
+  template <class T> struct hash;
+  template <class... Types> struct hash<variant<Types...>>;
+  template <> struct hash<monostate>;
+
+} // namespace std
+
+*/
+
+#include <cstddef>
+#include <exception>
+#include <functional>
+#include <initializer_list>
+#include <limits>
+#include <new>
+#include <type_traits>
+#include <utility>
+#include <Kokkos_Core.hpp>
+
+#include "config.hpp"
+#include "in_place.hpp"
+#include "lib.hpp"
+
+namespace mpark {
+
+  class bad_variant_access : public std::exception {
+    public:
+    virtual const char *what() const noexcept override { return "bad_variant_access"; }
+  };
+
+  [[noreturn]] KOKKOS_INLINE_FUNCTION void throw_bad_variant_access() {
+#if defined(MPARK_EXCEPTIONS)
+    throw bad_variant_access{};
+#elif  defined(__CUDA_ARCH__) ||       \
+    defined(__HIP_DEVICE_COMPILE__) || \
+    defined(__SYCL_DEVICE_ONLY__)
+    __assert_fail(nullptr, nullptr, 0, nullptr);
+#else
+    std::terminate();
+#endif
+    MPARK_BUILTIN_UNREACHABLE;
+  }
```

Comment:
Maybe the macros `KOKKOS_IF_ON_DEVICE(...)` and `KOKKOS_IF_ON_HOST(...)` could be used instead of checking for each specific device (they do the same check in [their macro](https://github.com/kokkos/kokkos/blob/1ca195b9e66a57770f91df5f0a5bce0d016a6815/core/src/Kokkos_Macros.hpp#L569-L583))
```suggestion
  [[noreturn]] KOKKOS_INLINE_FUNCTION void throw_bad_variant_access() {
#if defined(MPARK_EXCEPTIONS)
    throw bad_variant_access{};
#else
    KOKKOS_IF_ON_HOST((std::terminate();))
    KOKKOS_IF_ON_DEVICE((__assert_fail(nullptr, nullptr, 0, nullptr);))
#endif
  }
```